core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -212,7 +213,8 @@ class TransactionCoordinator(brokerId: Int,
           // particular, if fencing the current producer exhausts the available epochs for the current producerId,
           // then when the client retries, we will generate a new producerId.
           Right(coordinatorEpoch, txnMetadata.prepareFenceProducerEpoch())
-        case Dead =>
+
+        case Dead | PrepareEpochFence =>
```

Comment:
I'm not sure I understand why this is an illegal state. Isn't it possible for a transaction to be aborted at about the same time that a new producer sends InitProducerId?

Reply:
When we fence the epoch, we immediately set the `txnMetadata.pendingState` to `PrepareEpochFence`. This locks all other requests, since any new request which comes in when the pending state is defined will be rejected with the `CONCURRENT_TRANSACTIONS` error code.

Once we are in this pending state, we immediately begin ending the transaction, in which case the pending state becomes `PrepareAbort`. We then follow the normal state transitions for the end transaction process, which ultimately ends in `CompleteAbort`. 

So at no time do we ever complete the transition to `PrepareEpochFence`. So if the transaciton metadata is in this state, it is always illegal.

Reply:
@guozhangwang explained offline. The `PrepareEpochFence` is special in the sense that you can't actually transition to it. This is quite confusing, but I guess the check makes sense.